# notes_on_bioinfo_basics

Here are some notes I made after my thesis project for future me and anyone who are interested in learning the basics of bioinformatics! :-D 

These notes are intended for someone with basic understanding of `bash`/`R` and wish to learn more about tools or scripts that might be useful during bioinformatics analysis. I would be very grateful if you can let me know if there are any mistakes, so that I can improve my skills with you! l am very happy to know if there is anything I could add to make it more comprehensive.

**Things planning to add**
- Bash, HPC, and Conda (Section 1)
- Bulk RNA sequencing for differential gene expression analysis (Section 2; second part in progress...ï¼‰
- ChIP-Seq, ATAC-seq, Cut&Tag/Run (in progress...)
- Hi-C, Micro-C (in progress)

**What are covered in each section?**
- **Section 1** is entirely on `bash`. Here we discussed some common commands useful for navigating in the Linux system and doing things to files. Topics covered: installation for Windows system; useful `bash` commands, requesting resources and submitting jobs with PBS; creating and managing Conda environment.
- **Section 2** is also on `bash`. Here we discussed the overview of bulk RNAseq for DGE analysis, quality checking, and alignment. Topics covered: quality checks before alignment, adapter trimming, alignment using `STAR`, quality checks after alignment using `qualimap` and `RSeQC`.
- In later sections there might be scripts from `R` as well.
